libmproxy/web/app.py

```diff
@@ -25,6 +27,41 @@ def set_default_headers(self):
             "style-src   'self' 'unsafe-inline'"
         )
 
+    def set_auth_headers(self):
+        self.set_status(401)
+        self.set_header('WWW-Authenticate', 'Basic realm=MITMWeb')
+        self._transforms = []
+        self.finish()
+
+    def initialize(self, **kwargs):
+        self.singleuser = kwargs.get("singleuser")
+        self.htpasswd = kwargs.get("htpasswd")
+
+    def prepare(self):
+        if self.singleuser or self.htpasswd:
+            auth_header = self.request.headers.get('Authorization')
+            if auth_header is None or not auth_header.startswith('Basic '):
+                self.set_auth_headers()
+            else:
+                self.auth_decoded = base64.decodestring(auth_header[6:])
+                self.username, self.password = self.auth_decoded.split(':', 2)
+                if self.singleuser:
+                    if len(self.singleuser.split(':')) != 2:
+                        raise APIError(400, "Invalid single-user specification. Please use the format username:password")
+                    username, password = self.singleuser.split(':')
+                    password_manager = authentication.PassManSingleUser(username, password)
+                    if not password_manager.test(self.username, self.password):
+                        self.set_auth_headers()
+                        raise APIError(401, "Invalid username or password.")
+                elif self.htpasswd:
+                    try:
+                        password_manager = authentication.PassManHtpasswd(self.htpasswd)
+                    except ValueError as v:
+                        raise APIError(500, "%s. Please contact your systems administrator." % v.message)
+                    if not password_manager.test(self.username, self.password):
+                        self.set_auth_headers()
+                        raise APIError(401, "Invalid username or password.")
```

Comment:
This looks like there's quite a bit of duplication here. It'd be useful if we could just pass a password_manager to whatever authentication thingie we're using. The `password_manager` could be initialized when we parse the command line parameters.